pil-analyzer/tests/condenser.rs

```diff
@@ -471,3 +484,158 @@ fn intermediate_arr_wrong_length() {
+            });
+        };
+"#;
+    let analyzed = analyze_string::<GoldilocksField>(input);
+    let expected = r#"namespace std::prover;
+    let eval = 8;
+namespace std::convert;
+    let fe = 9;
+namespace N(16);
+    col witness x;
+    std::prelude::set_hint(N::x, {
+        let r = 9;
+        let k = [-2];
+        let q = std::prelude::false;
+        let b = {
+            let s = "";
+            (|| "" == s)
+        };
+        (|i| { N::y(1, i, (|| if b() && q { 9 + r } else { std::convert::fe::<int>(k[0]) })) })
+    });
+    let y: fe, int, (-> fe) -> std::prelude::Query = (|a, b, c| std::prelude::Query::Hint(a + c()));
+"#;
+    assert_eq!(analyzed.to_string(), expected);
+}
+
+#[test]
+fn prover_functions() {
+    let input = "
+    namespace std::convert;
+        let fe = 8;
+    namespace std::prover;
+        let provide_value = 9;
+    namespace N(16);
+        let x;
+        let y;
+        x = y;
+        query |i| {
+            std::prover::provide_value(x, i, std::convert::fe(i % 2));
+            std::prover::provide_value(y, i, std::convert::fe(i % 2));
+        };
+    ";
+    let analyzed = analyze_string::<GoldilocksField>(input);
+    let expected = r#"namespace std::convert;
+    let fe = 8;
+namespace std::prover;
+    let provide_value = 9;
+namespace N(16);
+    col witness x;
+    col witness y;
+    N::x = N::y;
+    (query |i| {
+        std::prover::provide_value(N::x, i, std::convert::fe::<int>(i % 2));
+        std::prover::provide_value(N::y, i, std::convert::fe::<int>(i % 2));
+
+    });
+"#;
+    assert_eq!(analyzed.to_string(), expected);
+}
+
+#[test]
+fn prover_functions_dynamic() {
+    let input = "
+    namespace std::convert;
+        let fe = 8;
+    namespace std::prover;
+        let provide_value = 9;
+    namespace N(16);
+        let gen = constr || {
+            let x;
+            let y;
+            x = y;
+            query |i| {
+                std::prover::provide_value(x, i, std::convert::fe(i % 2));
+                std::prover::provide_value(y, i, std::convert::fe(i % 2));
+            };
+        };
+        gen();
+    ";
+    let analyzed = analyze_string::<GoldilocksField>(input);
+    let expected = r#"namespace std::convert;
+    let fe = 8;
+namespace std::prover;
+    let provide_value = 9;
+namespace N(16);
+    let gen: -> () = (constr || {
+        let x: col;
+        let y: col;
+        x = y;
+        (query |i| {
+            std::prover::provide_value(x, i, std::convert::fe::<int>(i % 2));
+            std::prover::provide_value(y, i, std::convert::fe::<int>(i % 2));
+
+        });
+
+    });
+    col witness x;
+    col witness y;
+    N::x = N::y;
+    {
+        let x = N::x;
+        let y = N::y;
+        (query |i| {
+            std::prover::provide_value(x, i, std::convert::fe::<int>(i % 2));
+            std::prover::provide_value(y, i, std::convert::fe::<int>(i % 2));
+
+        })
+    };
+"#;
+    assert_eq!(analyzed.to_string(), expected);
+}
```

Comment:
What does "dynamic" mean here? What if there are `x` and `y` already created outside `gen()`, will `gen()` dynamically generated two different column names?

Reply:
Dynamic means that in line 600, a column with the proposed name `x` is created each time the funciton `gen` is called. The name is disambiguated by adding a suitable prefix if the name is already taken.

Reply:
Alrighty. Thanks